Tools to generate DL3 files with IRFs produced using pyirf

chaimain opened the pull request:
New draft PR forwarded from #511 and incorporating all the previous comments to make it cleaner.

I have separated the script of creating DL3 files into -
1. Generating IRFs (point-like Effective Area and Energy Dispersion) from MC DL2 files, using latest [pyirf](https://github.com/cta-observatory/pyirf/) v0.3.0.
2. Generating event lists from real LST data DL2 files, and then have an option to include or exclude existing IRFs, to generate DL3 file.

A new sub-module is created called irf, which contains functions to generate event lists and DL3 index files (HDU_INDEX and OBS_INDEX).
Currently, I use fixed selection cuts, written in data/data_selection_cuts.json.
Unit tests have been added for the changes made.

chaimain commented on lstchain/io/io.py:
```diff
@@ -1,26 +1,32 @@
 import h5py
 import numpy as np
-from astropy.table import Table, vstack
+import pandas as pd
```
Just made some rearrangement of the imports.

maxnoe commented on lstchain/io/io.py:
```diff
@@ -1018,7 +1025,29 @@ def read_dl2_to_pyirf(filename):
+    }
+    #TODO: Update other parameters with keys
+    data = pd.read_hdf(filename, key = dl2_params_lstcam_key).rename(columns=name_mapping)
+    data = table.QTable.from_pandas(data)
```
units?

chaimain replied:
I will add them now. Was earlier adding the units in the main scripts, but will update them.

maxnoe commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python3
```
Since this is new code, I would suggest to use a ctapipe Tool for this script.

maxnoe commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,293 @@
+        # Event list
+        t_events = Table(
+                {'OBS_ID':[event_table.meta['OBS_ID']],'HDU_TYPE':['events'],'HDU_CLASS':['events'],
+                    'HDU_CLASS2':[''],'HDU_CLASS3':[''],'HDU_CLASS4':[''],'FILE_DIR':[''],'FILE_NAME': [file],
```
'events' should be upper case here and I don't think you need to add empty `HDU_CLASSX` header entries

maxnoe commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,293 @@
+    event_table = QTable(
+            {
+                "EVENT_ID" : u.Quantity(data['event_id']),
+                "TIME" : u.Quantity(data['dragon_time'] * u.s),
```
if data is a QTable, shouldn't it already have units?

maxnoe commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,293 @@
+            {
+                "EVENT_ID" : u.Quantity(data['event_id']),
+                "TIME" : u.Quantity(data['dragon_time'] * u.s),
+                "RA" : u.Quantity(coord_icrs.ra.deg * u.deg),
```
Unnecessary, just use `coord_icrs.ra.to(u.deg)`.

maxnoe commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,293 @@
+                "TIME" : u.Quantity(data['dragon_time'] * u.s),
+                "RA" : u.Quantity(coord_icrs.ra.deg * u.deg),
+                "DEC" : u.Quantity(coord_icrs.dec.deg * u.deg),
+                "ENERGY" : u.Quantity(data['reco_energy'] * u.TeV)
```
same as above

maxnoe commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,293 @@
+
+    ev_header["RA_PNT"]=float(coord_pointing.icrs.ra.deg)
+    ev_header["DEC_PNT"]=float(coord_pointing.icrs.dec.deg)
+    ev_header["ALT_PNT"]=round(np.rad2deg(data['alt_tel'][0]),6)
```
I would not give `ALT_PNT` for a tracking observation. At the very least, it should be the `mean` alt pointing, not the first entry.

maxnoe commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,214 @@
+        p["events"]["theta"] = calculate_theta(
+            p["events"],
+            assumed_source_az=p["events"]["true_az"],
+            assumed_source_alt=p["events"]["true_alt"],
```
this will be wrong for protons / electrons

maxnoe replied:
This should be the wobble position or the center of the camera depending on if you do off-axis or on-axis observations.

maxnoe commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,214 @@
+
+    gammas = particles["gamma"]["events"]
+    # selecting 1 tel_id as the data is collected from only 1 for now
+    gammas = gammas[gammas["tel_id"]==1]
```
What if people want to use this script for MAGIC / LST observations. Shouldn't this at least be configurable?

maxnoe commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,214 @@
+    edisp = energy_dispersion(gammas[gammas["selected"]], true_energy_bins, fov_offset_bins, migration_bins)
+    hdus.append(create_energy_dispersion_hdu(edisp,true_energy_bins, migration_bins, fov_offset_bins, extname = "ENERGY DISPERSION"))
+
+    output_file = (output_dir/"irf.fits.gz")
```
no need for parentheses here

maxnoe commented on lstchain/io/io.py:
```diff
@@ -1,26 +1,32 @@
 from ctapipe.io import HDF5TableWriter
+from ctapipe.tools.stage1 import Stage1ProcessorTool
+from ctapipe.instrument import (OpticsDescription, CameraGeometry,
+		 CameraDescription, CameraReadout, \
```
you don't need line contiuations inside parentheses.

maxnoe commented on lstchain/scripts/lstchain_dl2_to_dl3.py:
```diff
@@ -109,7 +109,7 @@ def main():
     # Assuming the filename to be 'dl2_LST-1.Run#####_*.h5'
```
There are functions for this in lstchain:

```
from lstchain.paths import run_info_from_filename
run_info = run_info_from_filename(filename)
```

chaimain replied:
Thanks for pointing this out!

maxnoe commented on lstchain/io/io.py:
```diff
@@ -1018,7 +1025,46 @@ def read_dl2_to_pyirf(filename):
+        'pointing_az': u.rad,
+        'reco_alt': u.rad,
+        'reco_az': u.rad,
+		'dragon_time': u.s,
```
Indention here is mixed tabs/spaces. This should not happen with any good editor.

Python should be indented only by 4 spaces / level.

chaimain replied:
Corrected. Thanks!

maxnoe commented on lstchain/scripts/lstchain_dl2_to_dl3.py:
```diff
@@ -0,0 +1,156 @@
+parser.add_argument('--source-name', '-s', type=str,
+                    dest='source_name',
+                    help='Name of the source',
+                    default='Crab', required=True
```
Default and required are mutually exclusive. What is it?

chaimain replied:
Ah yes! I will make default None, to have this argument required.

maxnoe replied:
No, you just don't need a default if `required=True`.

chaimain replied:
Or, I can remove the default altogether for this and the obs_mode argument. WIll that be better?

maxnoe replied:
Yes. In python you dont need to give arguments when they have a default.

maxnoe commented on lstchain/scripts/lstchain_dl2_to_dl3.py:
```diff
@@ -0,0 +1,156 @@
+parser.add_argument('--obs-mode', '-mode', type=str,
+                    dest='obs_mode',
+                    help='Observation mode: ON, OFF or WOBBLE',
+                    default=None, required=True
```
same as above

maxnoe commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,177 @@
+                    default=None, required=False
+                    )
+
+parser.add_argument('--pnt-like', '-pnt', action='store',
```
use `action='store_true'` for boolean flags.

chaimain replied:
Ok thanks!

maxnoe commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,177 @@
+parser.add_argument('--pnt-like', '-pnt', action='store',
+                    type=lambda x: bool(strtobool(x)), dest='irf_type',
+                    help='True for point-like IRF, False for Full Enclosure',
+                    default=None, required=False
```
These two arguments are redundant.

Also, why are you storing this in `irf_type`? an `irf_type` variable that is True for point-like and False for FE is very confusing. Better just use `--point-like`.

chaimain replied:
OK, thanks!

maxnoe commented:
Since these are new scripts, it would be better to implement them using `ctapipe.core.Tool` not using argparse.

chaimain commented:
> Since these are new scripts, it would be better to implement them using `ctapipe.core.Tool` not using argparse.

Ok, I will do so.

vuillaut commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,323 @@
+    pointing_alt = data['pointing_alt']
+    pointing_az = data['pointing_az']
+
+    coord = camera_to_altaz(pos_x = pos_x, pos_y=pos_y, focal = focal,
+                    pointing_alt = pointing_alt, pointing_az = pointing_az,
+                    obstime = time)
+    coord_pointing = camera_to_altaz(pos_x = 0 * u.m, pos_y=0 * u.m, focal = focal,
+                pointing_alt = pointing_alt[0], pointing_az = pointing_az[0],
+                obstime = time[0])
+
+    object_radec=SkyCoord.from_name(source_name)
+
+    ##########################################################################
+    ### Event table
+    event_table = QTable(
+            {
+                "EVENT_ID" : u.Quantity(data['event_id']),
+                "TIME" : u.Quantity(data['dragon_time']),
+                "RA" : u.Quantity(coord.icrs.ra.to(u.deg)),
+                "DEC" : u.Quantity(coord.icrs.dec.to(u.deg)),
+                "ENERGY" : u.Quantity(data['reco_energy'])
+            }
+        )
```
I don't understand this part
The DL2 data must include the reconstructed position in the sky already (`reco_alt` and `reco_az`).
Is there a reason to reconstruct these coordinates again?

chaimain replied:
I have amended this.

chaimain commented:
Sorry for the long delay. I have added 2 Tools to create IRF file and DL3 file, which are a bit simple. Kindly let me know of any comments and feedback for improving the scripts/Tools.

LukasNickel commented on lstchain/tools/lstchain_create_irf_files.py:
```diff
@@ -0,0 +1,189 @@
+        else:
+            self.cuts = read_configuration_file(self.config_file)
+
+        if str(self.input_gamma_dl2) is not None:
```
Why is that string conversion necessary?

maxnoe replied:
This if is never true because str will never return `None`

chaimain replied:
I incorrectly assumed that it was working. On this, I would like to ask if we should have a separate argument for diffuse gamma or not. One idea I have is to use a boolean as an argument, just like for point-like IRF marker, and to include these markers in the config file itself.
Kindly let me know if you have any suggestions for including it in this PR.

maxnoe replied:
You should be able to deduce from the simulation header (view_cone) what kind of file it is. You shouldn't need to specify it as a cli argument.

chaimain replied:
Amended the files accordingly.

LukasNickel commented on lstchain/tools/lstchain_create_irf_files.py:
```diff
@@ -0,0 +1,189 @@
+        reco_energy_bins = create_bins_per_decade(0.01 * u.TeV, 100 * u.TeV, 5.5)
+
+        #TODO: Generalize FoV offset binning
+        if str(self.input_gamma_dl2) is None:
```
same here

maxnoe replied:
This if is also never True

chaimain replied:
Corrected here as well.

LukasNickel commented on environment.yml:
```diff
@@ -7,7 +7,7 @@ dependencies:
   - pip
   - ctapipe=0.8.0
-  - gammapy>=0.17
+  - gammapy~=0.17
```
What prevents the use of gammapy 0.18?

maxnoe replied:
This will use any gammapy 0.x version larger than 0.17

LukasNickel replied:
Ah right. I thought its any 0.17.x version

LukasNickel commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,354 @@
+            t_edisp['HDU_NAME'] = ['ENERGY DISPERSION']
+
+            hdu_tables.append(t_edisp)
+        except:
```
Those excepts are very general and the error message might not always be appropriate

chaimain replied:
Agreed. I have updated them. Let me know if it could be further improved

LukasNickel commented on lstchain/scripts/lstchain_create_dl3_index_files.py:
```diff
@@ -0,0 +1,68 @@
+                    default=None, required=True
+                    )
+
+parser.add_argument('--num-files', '-n', type=int,
```
This is supposed to take the n files from the folder and process them?
A more general way, that only needs one argument,  would be to provide an expression.
So instead of `path/to/files` you could give something like `path/to/files/run00[3-4]*`
This could be fed into pathlibs glob or the python standard library glob to create the filelist.

chaimain replied:
Thank you for this suggestion! I have implemented it and now the script is much crisper!

LukasNickel commented on lstchain/io/io.py:
```diff
@@ -1018,7 +1025,46 @@ def read_dl2_to_pyirf(filename):
     events = pd.read_hdf(filename, key=dl2_params_lstcam_key).rename(columns=name_mapping)
     events = table.QTable.from_pandas(events)
 
+    #Make the columns as Quantity
```
Slightly Offtopic:
Are there any code style guidelines enforced in lstchain?
Ctapipe has a black pre commit hook, that would fix some of the style things (Whitespace after #).

chaimain replied:
Yes, lstchain follows the same code style guidelines as ctapipe.
I had forgotten to take it into account while writing my scripts. Thank you for pointing it out! I will soon update accordingly.

LukasNickel commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,354 @@
+
+    return
+
+def create_event_list(data, run_number, source_name):
```
A lot of hard coded things.
Might be necessary right now, but we should keep track of those.

chaimain replied:
Indeed!

maxnoe commented on lstchain/tools/lstchain_create_dl3_file.py:
```diff
@@ -0,0 +1,145 @@
+
+        if self.add_irf:
+            irf = fits.open(self.input_irf)
+            self.aeff2d = irf[1]
```
Just access by using extreme, hdulists allow both access by index and access by hdu name.

Access by index here must not be used.

LukasNickel replied:
Thats even better

chaimain replied:
Thank you for the suggestion @LukasNickel and @maxnoe. I have corrected it.

chaimain commented on lstchain/tools/lstchain_create_dl3_file.py:
```diff
@@ -0,0 +1,145 @@
+
+    def start(self):
+
+        # Add angular separation column, different from pyirf function, because
```
Yes, sorry, had forgotten about this line.

chaimain commented:
Thank you @LukasNickel and @maxnoe for your feedback! I have tried to accommodate all your feedback and suggestions. Kindly let me know on any further feedback.

vincentpoireau commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,170 @@
+$> python lstchain_mc_dl2_to_irf.py
+--input-file-gamma ./gamma/dl2_gamma_*.h5
+--output-dir ./IRFs/
+--pnt-like True
```
The option is point-like
```suggestion
--point-like True
```

chaimain replied:
Thank you, I am correcting these errors.

vincentpoireau commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,170 @@
+Usage:
+$> python lstchain_mc_dl2_to_irf.py
+--input-file-gamma ./gamma/dl2_gamma_*.h5
+--output-dir ./IRFs/
```
The option is output-irf-dir
```suggestion
--output-irf-dir ./IRFs/
```

vincentpoireau commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,170 @@
+                    default=None, required=False
+                    )
+
+parser.add_argument('--output-irf-dir', '-o', type=Path, dest='output_irf_dir',
```
No need for dest here and below (but above, yes, since input-file-gamma and gamma_file are different names)

chaimain replied:
Yes, for keeping the style the same, I have changed gamma_file to input_file_gamma.

vincentpoireau commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,170 @@
+        gammas["selected_tels"] = gammas["tel_id"] == i
+
+    gammas["selected_gh"] = gammas["gh_score"] > gh_cut
+    # irf_type = True for point like IRFs, False for Full Enclosure IRFs
```
```suggestion
    # point_like = True for point like IRFs, False for Full Enclosure IRFs
```

vincentpoireau commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,170 @@
+        else:
+            effective_area = effective_area_per_energy_and_fov(gammas[gammas["selected"]], mc_gamma["simulation_info"], true_energy_bins, fov_offset_bins)
+    # Adding a dimension for FoV offset for effective area
+    hdus.append(create_aeff2d_hdu(effective_area[..., np.newaxis],true_energy_bins, fov_offset_bins, extname = "EFFECTIVE AREA"))
```
Why do you need to add a dimension for the effective area? Is this related to a Gammapy bug? FYI, there is a discussion on this topic in the Slack of Gammapy dated of the 18-19 Jan 2021, with some ideas to deal with it.

chaimain replied:
Yes. It was related to an issue with gammapy. I will update with the discussion on Slack and correct this.

vincentpoireau commented on lstchain/scripts/lstchain_mc_dl2_to_irf.py:
```diff
@@ -0,0 +1,170 @@
+- Output: IRFs compiled in fits.gz format
+
+Usage:
+$> python lstchain_mc_dl2_to_irf.py
```
The name of the IRF is now hard coded in the file (irf.fits.gz). It would be great to have an option to set the name, eg --irf-file-name.

chaimain replied:
I think I assumed it to be the standard name used. I will check if gammapy has any issue with any other name.

vincentpoireau replied:
No, there are no issues with a change of name. I have to change the name to have a different name for diffuse or point-like MC. And this is very convenient since in your script lstchain_dl2_to_dl3.py, you have the option --input-irf-file which allows to give the name of the input irf file. I do that and encounter no problem. I had to rename by hand the fits file, but would be more convenient to do it with an option in the script lstchain_mc_dl2_to_irf.py (and better also when everything will be automatic).

chaimain replied:
Ok! I understand. I can change the output-irf-dir to output-irf-file, so that the user provides the name and path of the output file, like in lstchain_merge_hdf5_files.py. Later if there is a standardised nomenclature, it can be updated.

vincentpoireau reviewed (commented):
Hi Chaitanya. Here at LAPP we use your scripts to produce DL3 files. So as users, we have several comments. Thanks for this nice code! Vincent.

vincentpoireau commented on lstchain/data/data_selection_cuts.json:
```diff
@@ -0,0 +1,22 @@
+  },
+  "fixed_cuts":{
+    "gh_score": [0.6,1],
+    "theta_cut": {"value": 0.2, "unit": "deg"},
```
Why do you put 0.2? I though the theta cut was 0.14 as we discussed a few months ago? (but maybe I missed a point)

chaimain replied:
Yes, this I had changed for my own analysis, where I loosened the theta cut for the data selected. This is just a fixed user-dependent cut for now. When we include the sensitivity calculations with the scripts, it will be changed accordingly.

chaimain commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,345 @@
+            t_edisp['HDU_TYPE'] = ['edisp']
+            t_edisp['HDU_CLASS'] = ['edisp_2d']
+            t_edisp['HDU_CLASS2'] = ['EDISP']
+            t_edisp['HDU_CLASS3'] = ['POINT-LIKE']
```
Thank you for pointing it out! Yes, for now it is indeed hard coded. I will add the information of the HDUCLAS3 in the compressed IRF file and extract that information while adding it to the HDU index table's IRF sections

chaimain commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,345 @@
+                ).deg
+    # Assuming wobble offset is fixed to 0.4
+    if round(source_pointing_diff, 1) == 0.4:
+        mode = 'WOBBLE'
```
Yes. I will make a comment at the top of the script as well about this.

chaimain commented on lstchain/scripts/lstchain_dl2_to_dl3.py:
```diff
@@ -0,0 +1,138 @@
+# Required arguments
+parser.add_argument('--input-data', '-d', type=Path,
+                    dest='input_data',
+                    help='path to merged DL2 data HDF5 file',
```
Thank you for pointing this out! It is indeed true. I will amend the arguments accordingly in all the scripts.

chaimain commented on lstchain/scripts/lstchain_dl2_to_dl3.py:
```diff
@@ -0,0 +1,138 @@
+def main():
+
+    if not args.input_data.is_file():
+        log.error('Input Path does not exist or is not a file')
```
Ok. I will update it.

chaimain commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,345 @@
+            t_aeff['HDU_TYPE'] = ['aeff']
+            t_aeff['HDU_CLASS'] = ['aeff_2d']
+            t_aeff['HDU_CLASS2'] = ['AEFF']
+            t_aeff['HDU_CLASS4'] = ['AEFF_2D']
```
Here I am just copying the header values from the earlier energy dispersion, so the value will remain the same for all the IRFs hereafter. It is indeed hard-coded right now, but I think we can update it later. If you think it is essential to generalize it in the current script, let me know.

chaimain commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,345 @@
+        mode = 'ON'
+    else:
+        mode = 'ON-MISPOINTING' # Either this or modify the method of getting ON mode
+
```
Any suggestions for naming a 'mispointing' mode of observation? There is no set standard right now for this in GADF, but it will be useful later for science tools, and decide upon the 'quality' of the data.

vincentpoireau replied:
UNDETERMINED? Or UNDERTERMINED-POINTING?

chaimain replied:
I have kept it UNDETERMINED for now.

maxnoe commented:
@chaimain I am a bit confused, didn't you say in the last meeting that you are now using `ctapipe.core.Tool`? I still see scripts using `argparse` instead.

chaimain commented:
> @chaimain I am a bit confused, didn't you say in the last meeting that you are now using `ctapipe.core.Tool`? I still see scripts using `argparse` instead.

I have kept the older scripts as well for now. I am using the new tools in `lstchain/tools`. Shall I remove the older scripts completely?

maxnoe commented:
@chaimain it does not make sense to add duplicated functionality.

chaimain commented:
Ok. I will remove them now.

chaimain commented:
@maxnoe we need tests for running over the tools as well. And as we have some new tools coming as well, should I created a test for the tools in this PR itself, or it can be done in a separate PR?

maxnoe commented:
You should add tests for new code together with the code.

vincentpoireau commented on lstchain/tools/lstchain_create_irf_files.py:
```diff
@@ -0,0 +1,214 @@
+            gammas["selected_tels"] = gammas["tel_id"] == i
+
+        gammas["selected_gh"] = gammas["gh_score"] > gh_cut
+        # irf_type = True for point like IRFs, False for Full Enclosure IRFs
```
```suggestion
        # point_like = True for point like IRFs, False for Full Enclosure IRFs
```

vincentpoireau commented on lstchain/tools/lstchain_create_dl3_file.py:
```diff
@@ -0,0 +1,172 @@
+
+        self.data = self.data[self.data["reco_source_fov_offset"] < u.Quantity(
+                                        **self.cuts["fixed_cuts"]["source_fov_offset"])]
+        self.log.info("Gemerating Event List")
```
```suggestion
        self.log.info("Generating Event List")
```

vincentpoireau commented:
I ran with the lstchain/tools script to produce IRF and DL3, and this works fine. However I wonder why the script scripts/lstchain_create_dl3_index_files.py is not also in the lstchain/tools? (there may be good reasons) This script is used after the creation of DL3 to create the files hdu-index.fits.gz and obs-index.fits.gz.

vuillaut commented on lstchain/io/io.py:
```diff
@@ -1018,7 +1025,46 @@ def read_dl2_to_pyirf(filename):
+        'reco_alt': u.rad,
+        'reco_az': u.rad,
+        'dragon_time': u.s,
+        'reco_src_x': u.m,
```
should be removed, theses are not used

chaimain commented:
I have made the script to create HDU index files into a tool and added new unit tests for tools and the HDU functions used in those tools in lstchain/irf/tests/test_hdus.py.
I have also updated the IRF tool to also generate (preliminary) Full Enclosure IRFs.
Also in the IRF tool, the output IRF compressed FITS file can be named by the user.

chaimain commented:
Also, as I understand, we do not have any tests for Tools yet for lstchain (including the other open PRs) and we will need to update even this one, once we update the ctapipe dependency. So, please let me know if these tests can be further improved so that the other Tools in the open PRs can also be updated to a standard 'lstchain format'

maxnoe commented on lstchain/tools/lstchain_create_dl3_file.py:
```diff
@@ -1,6 +1,14 @@
 Create DL3 FITS file from given data DL2 file,
 selection cuts and/or IRF FITS files
+
+Simple usage with argument aliases and default config file for selection cuts:
```
Put this into the description of the tool so people can actually see this when typing `--help`

chaimain replied:
Ok, I will do so.

maxnoe replied:
To have it here, you could actually also do `description = __doc__`

maxnoe commented on lstchain/tools/lstchain_create_dl3_index_files.py:
```diff
@@ -1,6 +1,11 @@
 Create HDU index files for HDU tables and Obs tables,
 from a given path of DL3 files and a glob pattern to select DL3 files
+
+Simple usage with argument aliases:
```
Same as above

maxnoe commented on lstchain/tools/lstchain_create_irf_files.py:
```diff
@@ -2,6 +2,12 @@
 MC gamma files can be point-like or diffuse
 IRFs can be point-like or Full Enclosure
+
+Simple usage with argument aliases and default config file for selection cuts:
```
Same as above

maxnoe commented on lstchain/irf/hdu_table.py:
```diff
@@ -0,0 +1,345 @@
+            t_bkg['HDU_TYPE'] = ['bkg']
+            t_bkg['HDU_CLASS'] = ['bkg_2d']
+            t_bkg['HDU_CLASS2'] = ['BKG']
+            t_bkg['HDU_CLASS4'] = ['BKG_2D']
```
background 2d is available in `pyirf`. Why not use it?

maxnoe replied:
https://cta-observatory.github.io/pyirf/api/pyirf.io.create_background_2d_hdu.html#pyirf.io.create_background_2d_hdu

chaimain replied:
I have added it now.

chaimain commented:
I have updated the IRF Tools with pyirf's Background 2D HDU and added spectral weighting option for the MC data using pyirf's spectral functions (can update later with using from lstchain.spectra). Diffuse MC gamma uses Proton spectrum.

There is still the option to just produce Effective Area and Energy Dispersion IRFs with just providing MC gamma file as the argument for the IRF Tool, in case the user wants to use a Background model only defined from real data.

I believe, the PR is ready to be merged into the master. If anybody has any other urgent feedback to be added to the Tools through this PR, kindly let me know.

morcuended commented on lstchain/tools/lstchain_create_irf_files.py:
```diff
@@ -0,0 +1,359 @@
+        self.effective_area = None
+        self.edisp = None
+        self.backgroud = None
+        # self.psf = None
```
why is the PSF part commented out?

chaimain replied:
I have commented out PSF in this PR, as the Tools are currently used mainly for producing point-like IRFs, with the option to later complete it with full enclosure IRFs as well when the need arises.

chaimain commented:
The CI tests are failing for running 'download_test_data.sh', which only the main repository has access to. I will close this PR and reopen a new PR from a branch of the repository to master.

I will add the PSF IRF with a different PR after the DL3 tools are merged to master, and any other corrections to be made, can be done in such separate PRs.
